File: IT/Rete e web/Esercizi/C++/esercizio30/esercizio30.cpp
/*
file esercizio30.cpp
compilare con: g++ -std=c++11 esercizio30.cpp mongoose.c [-lwsock32]
*/

#include <string>
#include "mongoose.h"

using std::string;
using std::stoi;
using std::stod;
using std::to_string;

static void httpHandler(struct mg_connection* conn, int event, void* ev_data, void* fn_data) {
	if (event == MG_EV_HTTP_MSG) {
		struct mg_http_message *hm = (struct mg_http_message*) ev_data;

		if (mg_http_match_uri(hm, "/beratio")) {
			char loan[20];
			mg_http_get_var(&hm->body, "loan", loan, sizeof(loan));

			char debts[20];
			mg_http_get_var(&hm->body, "debts", debts, sizeof(debts));

			char nMonths[20];
			mg_http_get_var(&hm->body, "nMonths", nMonths, sizeof(nMonths));

			char income[20];
			mg_http_get_var(&hm->body, "income", income, sizeof(income));

			double _loan = stod(loan);
			double taeg = (_loan / 100) * 6.5;
			double _debts = stod(debts);
			int _nMonths = stoi(nMonths);
			double _income = stod(income);

			// calcola la rata mensile (incluso il TAEG): 
			double payment = ((_loan + taeg) / _nMonths);

			char buf[500];
			snprintf(buf,
					 sizeof(buf), 
					 "<!DOCTYPE html>"
					 "<html>"
					   "<body style=\"text-align: center\">"
					     "<div style=\"font-style: italic;\">Riassunto</div><br/>"
					     "<div>Prestito richiesto: %s euro</div>"
					     "<div>Reddito mensile: %s euro</div>"
					     "<div>Totale debiti pregressi: %s euro mensili</div>"
					     "<div>Numero rate desiderato: %s </div><br/>"
					     "<div style=\"font-weight: bold\">Importo rata mensile: %.2f </div>"
					     "<div>(Alla rata si applica un tasso di interesse annuo globale del 6.5%%%% annuo)</div><br/>",
					 loan,
					 income,
					 debts,
					 nMonths,
					 payment);

			double beratio = ((payment + _debts) / _income) * 100;

			if (beratio <= 36) {
				snprintf(buf + strlen(buf),
						 120,
						 "<div style=\"color:green;\">Rapporto debiti / reddito uguale a %.2f%. Possiamo concedere il prestito.</div>",
						 beratio);
			}
			else {
				snprintf(buf + strlen(buf),
						 120,
						 "<div style=\"color:red;\">Rapporto debiti / reddito eccessivo (%.2f%). Impossibile concedere il prestito.</div>",
						 beratio);
			}

			snprintf(buf + strlen(buf), 15, "</body></html>");

			mg_http_reply(conn, 200, "Content-Type: text/html; charset=utf-8\n", buf);
		}
		else {
			struct mg_http_serve_opts config = { ".", "#.html" };
			mg_http_serve_dir(conn, hm, &config);
		}
	}
}

int main()
{
	struct mg_mgr eventManager;
	struct mg_connection *conn;

	mg_mgr_init(&eventManager);

	mg_http_listen(&eventManager, "http://localhost:8080", httpHandler, &eventManager);

	while (true) {
		mg_mgr_poll(&eventManager, 1000);
	}

	mg_mgr_free(&eventManager);
}

File: IT/2/Rete e web/Esercizi/Python/esercizio30/esercizio30.py
# file esercizio30.py

import cherrypy

class Esercizio30(object):
	@cherrypy.expose
	def index(self):
		return open("index.html")

	@cherrypy.expose
	def beratio(self, loan, debts, income, nMonths):
		if loan and income:
			try:
				_loan = float(loan)
				_debts = float(debts) if debts else 0
				_nMonths = int(nMonths)
				_income = float(income)

				html = "<!DOCTYPE html><html><body style=\"text-align: center\"><div style=\"font-style: italic;\">Riassunto</div><br/>"
				html += "<div>Prestito richiesto: {} euro</div>".format(loan) 
				html += "<div>Reddito mensile: {} euro</div>".format(income)
				html += "<div>Totale debiti pregressi: {} euro mensili</div>".format(debts)
				html += "<div>Numero rate desiderato: {}</div><br/>".format(nMonths)

				if _loan < 500 or _loan > 30000:
					raise cherrypy.HTTPError(400, "L'importo del prestito richiesto dev'essere compreso tra 500 e 30000 euro!")

				# tasso d'interesse annuo del 6,5%
				taeg = (_loan / 100) * 6.5

				# calcola la rata mensile (incluso il TAEG): 
				payment = ((_loan + taeg) / _nMonths)

				html += "<div style=\"font-weight: bold\">Importo rata mensile: {0:.2f}</div>".format(payment)
				html += "<div>(Alla rata si applica un tasso di interesse annuo globale del 6.5% annuo.)</div><br/>"

				beratio = ((payment + _debts) / _income) * 100

				if beratio <= 36:
					html += "<div style=\"color:green;\">Rapporto debiti / reddito uguale a {0:.2f}%. Possiamo concedere il prestito.</div>".format(beratio)
				else:
					html += "<div style=\"color:red;\">Rapporto debiti / reddito eccessivo ({0:.2f}%). Impossibile concedere il prestito.</div>".format(beratio)

				html += "</body></html>"

				return html
			except ValueError as e:
				raise cherrypy.HTTPError(400, "Inserito un importo scorretto!")
		else:
			raise cherrypy.HTTPError(400, "Prestito e stipendio non possono essere nulli!")

if __name__ == "__main__":
	cherrypy.quickstart(Esercizio30())

